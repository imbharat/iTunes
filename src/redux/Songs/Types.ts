export type Song = {
    trackId: number,
    kind: string,
    artistName: string,
    collectionName: string,
    trackName: string,
    artworkUrl100: string
}

export type Filter = {
    term: string,
    page: number,
    country?: string,
    media?: string,
    entity?: string,
    attribute?: string,
    limit?: number,
    lang?: string
}

export type Songs = {
    loading: boolean,
    songs: Song[],
    isLast: boolean,
    filter: Filter,
    error: string
}

export type Action = {
    type: string,
    msg: string,
    payload: any
}