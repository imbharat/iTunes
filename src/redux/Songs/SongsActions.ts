import { fetchData } from "../../client/services/Search";
import { Filter, Songs } from "./Types";

export const ADD_SONGS = 'ADD_SONGS';
export const UPDATE_SONGS = 'UPDATE_SONGS';
export const LOAD_SONGS = 'LOAD_SONGS';
export const LOAD_FAILURE = 'LOAD_FAILURE';
export const UPDATE_FILTER = 'UPDATE_FILTER';

export const addSongs = (songs: Songs) => {
    return {
        type: ADD_SONGS,
        message: 'adding songs',
        payload: songs
    }
}

export const updateSongs = (songs: Songs) => {
    return {
        type: UPDATE_SONGS,
        message: 'updating songs',
        payload: songs
    }
}

export const loadSongs = () => {
    return {
        type: LOAD_SONGS,
        message: 'loading songs'
    }
}

export const loadFailure = (error: string) => {
    return {
        type: LOAD_FAILURE,
        message: 'error in loading',
        msg: error
    }
}

export const updateFilter = (filter: Filter) => {
    return {
        type: UPDATE_FILTER,
        message: 'updating filter',
        payload: filter
    }
}

export const updateFilterAndLoadSongs = (filter: Filter, refresh: boolean) => {
    return async function(dispatch: any){
        dispatch(updateFilter(filter));
        dispatch(getSongs(filter, refresh));
    }
}

export const getSongs = (filter: Filter, refresh: boolean) => {
    return async function(dispatch: any){
        dispatch(loadSongs());
        const response = await fetchData(filter);
        if(response?.data){
            const data = JSON.parse(response.data);
			refresh ? dispatch(updateSongs(data)) : dispatch(addSongs(data));
        }
		else
			dispatch(loadFailure('error in loading songs'));
    }
}