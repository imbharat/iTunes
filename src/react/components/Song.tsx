import React, { useState } from 'react'
import { Card, CardMedia, CardContent, Typography, Dialog, DialogContent } from '@mui/material';
import { StyledSong } from '../styled-components/StyledSong';
import { Song } from '../../redux/Songs/Types'

function Song({ song }: {song: Song}) {
    const {
        kind,
        artistName,
        collectionName,
        trackName,
        artworkUrl100
    } = song;

    const [viewDetails, setViewDetails] = useState(false);
    const handleClose = () => {
        setViewDetails(false);
    }

    return (
        <StyledSong>
            <Card onClick={() => setViewDetails(true)}>
                <CardMedia
                    component="img"
                    image={ artworkUrl100 }
                    alt={ trackName }
                />
                <CardContent>
                    <Typography gutterBottom variant="h4" component="div" className='trackN'>
                        { trackName }
                    </Typography>
                    <Typography gutterBottom variant="h6" component="div" className='collectionN'>
                        { collectionName ? ( collectionName ) : ''}
                    </Typography>
                    <Typography gutterBottom variant="h6" component="div" className='artistN'>
                        { artistName ? `-by: ${ artistName }` : ''}
                    </Typography>
                </CardContent>
            </Card>

            {/* <Dialog open={viewDetails} onClose={handleClose}>
                <DialogContent>
                    <Card>
                        <CardMedia
                            component="img"
                            image={ artworkUrl100 }
                            alt={ trackName }
                        />
                        <CardContent>
                            <Typography gutterBottom variant="h4" component="div" className='trackN'>
                                { trackName }
                            </Typography>
                            <Typography gutterBottom variant="h6" component="div" className='collectionN'>
                                { collectionName ? ( collectionName ) : ''}
                            </Typography>
                            <Typography gutterBottom variant="h6" component="div" className='artistN'>
                                { artistName ? `-by: ${ artistName }` : ''}
                            </Typography>
                        </CardContent>
                    </Card>
                </DialogContent>
            </Dialog> */}
        </StyledSong>
    )
}

export default Song