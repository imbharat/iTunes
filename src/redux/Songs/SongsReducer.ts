import { ADD_SONGS, LOAD_SONGS, LOAD_FAILURE, UPDATE_SONGS, UPDATE_FILTER } from "./SongsActions";
import { Songs, Action } from "./Types";

const initialSongs: Songs = {
    loading: false,
    songs: [],
    filter: {
        term: '',
        country: '',
        media: '',
        entity: '',
        attribute: '',
        limit: 50,
        lang: '',
        page: 0
    },
    isLast: true,
    error: ''
}

export const songsReducer = (state = initialSongs, action: Action) => {
    switch(action.type){
        case ADD_SONGS:
            return {
                ...state,
                songs: [
                    ...state.songs,
                    ...action.payload.data
                ],
                isLast: action.payload.isLast,
                error: ''
            }
        case UPDATE_SONGS:
            return {
                ...state,
                songs: action.payload.data,
                isLast: action.payload.isLast,
                error: ''
            }
        case LOAD_SONGS:
            return {
                ...state,
                loading: true,
                error: ''
            }
        case LOAD_FAILURE:
            const words = state.filter.term;
            return {
                ...initialSongs,
                filter: {
                    ...initialSongs.filter,
                    term: words
                },
                error: action.msg
            }
        case UPDATE_FILTER:
            return {
                ...state,
                filter: {
                    ...state.filter,
                    ...action.payload
                },
                error: ''
            }
        default:
            return {
                ...state
            }
    }
}