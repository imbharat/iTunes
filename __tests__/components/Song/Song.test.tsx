import React from 'react';
import renderer from 'react-test-renderer';
import Song from '../../../src/react/components/Song';
import { Song as SongObject } from '../../../src/redux/Songs/Types';

describe("Song Component Tests", () => {
    it('song component rendering', () => {
      //arrange
      const song: SongObject = {
        artistName: "Paul Gilbert",
        artworkUrl100: "https://is4-ssl.mzstatic.com/image/thumb/Music124/v4/88/aa/ed/88aaed59-2c30-27d4-7468-4b95ef5ee7bf/mzm.iwrtpfbi.jpg/100x100bb.jpg",
        collectionName: "Space Ship One",
        trackId: 1049049337,
        trackName: "G9",
        kind: "song"
      }
      //act
      const tree = renderer.create(<Song song={song}/>).toJSON();
      //assert
      expect(tree).toMatchSnapshot();
    });
});