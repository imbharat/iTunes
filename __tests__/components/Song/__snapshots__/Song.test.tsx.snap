// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Song Component Tests song component rendering 1`] = `
<div
  className="sc-bczRLJ gBgBPG"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    onClick={[Function]}
  >
    <img
      alt="G9"
      className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
      src="https://is4-ssl.mzstatic.com/image/thumb/Music124/v4/88/aa/ed/88aaed59-2c30-27d4-7468-4b95ef5ee7bf/mzm.iwrtpfbi.jpg/100x100bb.jpg"
    />
    <div
      className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom trackN css-1vw6mcs-MuiTypography-root"
      >
        G9
      </div>
      <div
        className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom collectionN css-18k87ye-MuiTypography-root"
      >
        Space Ship One
      </div>
      <div
        className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom artistN css-18k87ye-MuiTypography-root"
      >
        -by: Paul Gilbert
      </div>
    </div>
  </div>
</div>
`;
